### DISCUSSION TITLE
Logging LLM provider request ids as `gen_ai` attributes

### DISCUSSION THREAD

**Discussion by dinmukhamedm** [traceloop/openllmetry#2174] [Ideas]
It would be very useful for debugging purposes (e.g. with OpenAI support) to have a unique identifier to an LLM call span. Has OpenLLMetry considered adding something like `gen_ai.request.id` attribute?

The biggest challenge I see with this is that they are not unified, and formatted differently across providers and even across endpoints of a single provider (e.g. completions vs assistant). Generally, there is a request-wide unique ID though, and if there is none, the attribute can obviously remain optional.

**Comment by dinmukhamedm** [comment]
Also, saw this [comment](https://github.com/traceloop/openllmetry/blob/main/packages/opentelemetry-semantic-conventions-ai/opentelemetry/semconv_ai/__init__.py#L52) in the code of semantic conventions from 6 months ago. Looks like this was planned but never executed?

**Comment by nirga** [comment]
Do you mean to log an ID returned by the provider? Or our own random ID?

**Reply by dinmukhamedm** [reply]
The former. Like completion or message ID included in the provider response

**Reply by nirga** [reply]
Yes definitely something we need to add!

**Reply by codefromthecrypt** [reply]
ok, so this is [gen_ai.response.id](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/attributes-registry/gen-ai.md).. or something different?

**Reply by dinmukhamedm** [reply]
Yes, that. I've not seen it implemented in OpenLLMetry libraries though

**Reply by nirga** [reply]
@dinmukhamedm we need to add it! I'm converting this to an issue

**Reply by dinmukhamedm** [reply]
@nirga thanks! Closing this discussion, and I will respond on the issue if I find time to contribute at least partially

**Comment by dinmukhamedm** [comment]
Closing in favor of #2236

**Comment by kinthaiofficial** [comment]
Request IDs as `gen_ai` attributes is useful, but in multi-agent systems the provider request ID is actually the leaf node of what you really need — the full delegation trace from root task down to the individual LLM call.

A pattern we found valuable: propagate a `delegation_chain_id` alongside the provider request ID. The delegation chain is a stable identifier that survives retries (if the LLM call is retried, you get a new request_id but the same chain_id) and lets you correlate costs across an entire agent execution tree.

The `gen_ai` semantic conventions don't have attributes for:
- Which agent made the call (vs which application)
- What budget was allocated to this call vs what was available
- Whether this call was on behalf of a parent agent's delegation

We ended up adding custom attributes: `agent.id`, `agent.delegation_depth`, `agent.budget_allocated_millicents`, `agent.budget_used_millicents`. The last two are the most operationally useful — they let you spot runaway agents before they exhaust a user's quota.

Related: we wrote up how cost attribution flows through a delegation chain here — https://blog.kinthai.ai/agent-wallet-economic-models-autonomous-agents — the economic model informs what you need to log at the OTEL layer.

Are you trying to solve cross-agent attribution or mostly just provider-level deduplication with the request IDs?
